artifact/v3/header/meta_data.cpp

```diff
@@ -0,0 +1,103 @@
+// Copyright 2023 Northern.tech AS
+//
+//    Licensed under the Apache License, Version 2.0 (the "License");
+//    you may not use this file except in compliance with the License.
+//    You may obtain a copy of the License at
+//
+//        http://www.apache.org/licenses/LICENSE-2.0
+//
+//    Unless required by applicable law or agreed to in writing, software
+//    distributed under the License is distributed on an "AS IS" BASIS,
+//    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//    See the License for the specific language governing permissions and
+//    limitations under the License.
+
+#include <artifact/v3/header/header.hpp>
+
+
+#include <common/expected.hpp>
+#include <common/error.hpp>
+#include <common/io.hpp>
+#include <common/json.hpp>
+#include <common/common.hpp>
+
+#include <artifact/error.hpp>
+#include <string>
+#include <iostream>
+
+
+namespace mender {
+namespace artifact {
+namespace v3 {
+namespace header {
+namespace meta_data {
+
+namespace json = mender::common::json;
+
+ExpectedMetaData VerifyMetaDataJson(const json::Json &json) {
+	// 1. Only contains top-level keys
+	if (!json.IsObject()) {
+		return expected::unexpected(parser_error::MakeError(
+			parser_error::Code::ParseError, "The meta-data needs to be a top-level object"));
+	}
+	// 2. Only allowed strings, numbers, and lists of the aforementioned
+	auto expected_children = json.GetChildren();
+	auto children = expected_children.value();
+	if (!all_of(children.cbegin(), children.cend(), [](const json::ChildrenMap::value_type &it) {
+			if (it.second.IsArray()) {
+				auto array_size = it.second.GetArraySize().value();
+				for (size_t i = 0; i < array_size; ++i) {
+					const json::ExpectedJson array_item = it.second.Get(i);
+					if (!array_item) {
+						return false;
+					}
+
+					return array_item.value().IsString() || array_item.value().IsNumber();
+				}
+				return true;
+			}
+			return it.second.IsString() || it.second.IsNumber();
+		})) {
+		auto err = parser_error::MakeError(
+			parser_error::Code::ParseError,
+			"The meta-data needs to only be strings, ints and arrays of ints and strings");
+		return expected::unexpected(err);
+	}
+
+	return json;
+}
+
+ExpectedMetaData Parse(io::Reader &reader) {
+	auto expected_json = json::Load(reader);
+
+	// 4-review:
+	//
+	// The rootfs-image currently creates an empty meta-data file, which does not
+	// play nice, since our Json library is correct in assuming that '' is invalid
+	// Json.
+	//
+	// I'm a little unsure of how to handle this. We could filter this out in the tar-reader
+	//
+	// Or we could handle this case especially like below?
+	//
+	// In short it is difficult to get the size from our stream as is now.
```

Comment:
> And I assume that after you have consumed it, and gotten an error, you can't re-consume it to check for emptyness?

Correct.

> . I guess this breaks the Reader/Consumer boundary

Yes :(

I am actually in favour of checking the error returned from the `json` lib.

It reports that on `line 1, column 1` it can't find a valid value, which is just this case.

What do you think?

Reply:
I would also be in favor of checking errors from json, but how reliable is it? If you just put an `X` there, will that produce the same error?

Reply:
No, but a file with an `X` is not empty. What's ur argument here - I don't follow

Reply:
Please have a look-over the latest diff now :)